src/include/network/abstract_packet_writer.h

```diff
@@ -0,0 +1,399 @@
+      case NetworkMessageType::SSL_NO:
+        queue_.BufferWriteRawValue(type);
+        break;
+      default:
+        BeginPacket(type).EndPacket();
+    }
+  }
+
+  /**
+   * Begin writing a new packet. Caller can use other append methods to write
+   * contents to the packet. An explicit call to end packet must be made to
+   * make these writes valid.
+   * @param type
+   * @return self-reference for chaining
+   */
+  AbstractPacketWriter &BeginPacket(NetworkMessageType type) {
+    // No active packet being constructed
+    TERRIER_ASSERT(curr_packet_len_ == nullptr, "packet length is null");
+    if (type != NetworkMessageType::NO_HEADER) queue_.BufferWriteRawValue(type);
+    // Remember the size field since we will need to modify it as we go along.
+    // It is important that our size field is contiguous and not broken between
+    // two buffers.
+    queue_.BufferWriteRawValue<int32_t>(0, false);
+    WriteBuffer &tail = *(queue_.buffers_[queue_.buffers_.size() - 1]);
+    curr_packet_len_ = reinterpret_cast<uint32_t *>(&tail.buf_[tail.size_ - sizeof(int32_t)]);
+    return *this;
+  }
+
+  /**
+   * Append raw bytes from specified memory location into the write queue.
+   * There must be a packet active in the writer.
+   * @param src memory location to write from
+   * @param len number of bytes to write
+   * @return self-reference for chaining
+   */
+  AbstractPacketWriter &AppendRaw(const void *src, size_t len) {
+    TERRIER_ASSERT(curr_packet_len_ != nullptr, "packet length is null");
+    queue_.BufferWriteRaw(src, len);
+    // Add the size field to the len of the packet. Be mindful of byte
+    // ordering. We switch to network ordering only when the packet is finished
+    *curr_packet_len_ += static_cast<uint32_t>(len);
+    return *this;
+  }
+
+  /**
+   * Append a value onto the write queue. There must be a packet active in the
+   * writer. No byte order conversion is performed. It is up to the caller to
+   * do so if needed.
+   * @tparam T type of value to write
+   * @param val value to write
+   * @return self-reference for chaining
+   */
+  template <typename T>
+  AbstractPacketWriter &AppendRawValue(T val) {
+    return AppendRaw(&val, sizeof(T));
+  }
+
+  /**
+   * Append a value of specified length onto the write queue. (1, 2, 4, or 8
+   * bytes). It is assumed that these bytes need to be converted to network
+   * byte ordering.
+   * @tparam T type of value to read off. Has to be size 1, 2, 4, or 8.
+   * @param val value to write
+   * @return self-reference for chaining
+   */
+  template <typename T>
+  AbstractPacketWriter &AppendValue(T val) {
+    // We only want to allow for certain type sizes to be used
+    // After the static assert, the compiler should be smart enough to throw
+    // away the other cases and only leave the relevant return statement.
+    static_assert(sizeof(T) == 1 || sizeof(T) == 2 || sizeof(T) == 4 || sizeof(T) == 8, "Invalid size for integer");
```

Comment:
How is `static_assert` different than `TERRIER_ASSERT`?

Reply:
Seems like `static_assert` throws a compile-time error, `TERRIER_ASSERT` uses `assert` so it throws a run-time error